numpy/__init__.pyi

```diff
@@ -1231,3 +1231,114 @@ def amin(
     initial: _NumberLike = ...,
     where: _ArrayLikeBool = ...,
 ) -> Union[number, ndarray]: ...
+
+# TODO: `np.prod()``: For object arrays `initial` does not necasarily
+# have to be a numerical scalar.
+# The only requirement is that it is compatible
+# with the `.__mul__()` method(s) of the passed array's elements.
+
+# Note that the same situation holds for all wrappers around
+# `np.ufunc.reduce`, e.g. `np.sum()` (`.__add__()`).
+
+@overload
+def prod(
+    a: _Number,
+    axis: Optional[_ShapeLike] = ...,
+    dtype: DtypeLike = ...,
+    out: None = ...,
+    keepdims: bool = ...,
+    initial: _NumberLike = ...,
+    where: _ArrayLikeBool = ...,
+) -> _Number: ...
+@overload
+def prod(
+    a: ArrayLike,
+    axis: None = ...,
+    dtype: DtypeLike = ...,
+    out: None = ...,
+    keepdims: Literal[False] = ...,
+    initial: _NumberLike = ...,
+    where: _ArrayLikeBool = ...,
+) -> number: ...
```

Comment:
I am assuming we are allowing number subtypes for one but not the other why is that ?
In other words, why is there a difference in return types of the above two overloads ?

Reply:
The key difference is that the first one is a `TypeVar` while the latter is not.

To illustrate, in the first overload an arbitrary `number` subclass (denoted by `_Number`) 
is provided as input and a value of the same type is returned.

In the second overload we know that an instance of `np.int_` is returned, 
but there is a catch: `np.int_` is a platform-dependent alias for one of the `number` subclasses, 
which is not something we've been able to easily express as of yet.

https://github.com/numpy/numpy/blob/92665ab229a2426db33bcfaa65cd73493acbad95/numpy/__init__.pyi#L511-L517

As a workaround the latter case is now simply annotated as returning _a_ `number` instance and
while correct, this is obviously not as specific as it could be.

Reply:
okay got it, if my understanding is correct, the functionality is exactly the same if we use number of _Number here (second overload), but it is something we want to change in the future ?

Reply:
> ..., but it is something we want to change in the future ?

Correct, once we've dealt with these platform-dependent `generic` subclasses.

Reply:
Oh hang on, I'm actually confusing the issues pertaining `int_` and `number` here, sorry.

The problem with returning`number` is more straightforward.
As `ArrayLike` is not yet `Generic` with respect to the data type we don't have a way yet to correlate the input and output types. 
In the future we should be able to use expressions such as the one below (or at least something similar):
```python
>>> from typing import TypeVar
>>> import numpy as np
>>> from numpy.typing import ArrayLike

>>> T = TypeVar('T', bound=np.number)

>>> def func(a: ArrayLike[T]) -> T:
...     pass

```

If you're interested, see https://github.com/numpy/numpy/pull/16759 for more details.